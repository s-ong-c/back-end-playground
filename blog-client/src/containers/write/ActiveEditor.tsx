import * as React from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import { WriteMode } from '../../modules/write';
import EditorPanesContainer from './EditorPanesContainer';
import QuillEditorContainer from './QuillEditorContainer';

interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
export type ActiveEditorProps = OwnProps & StateProps & DispatchProps;

const mapStateToProps = ({ write }: RootState) => ({
  mode: write.mode,
});
const mapDispatchToProps = {};

const ActiveEditor: React.FC<ActiveEditorProps> = ({ mode }) => {
  return mode === WriteMode.MARKDOWN ? (
    <EditorPanesContainer />
  ) : (
    <QuillEditorContainer />
  );
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(ActiveEditor);
