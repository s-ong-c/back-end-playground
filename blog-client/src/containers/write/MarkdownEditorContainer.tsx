import * as React from 'react';
import MarkdownEditor from '../../components/write/MarkdownEditor';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import {
  changeMarkdown,
  changeTitle,
  setHtml,
  convertEditorMode,
  WriteMode,
} from '../../modules/write';

import remark from 'remark';
import htmlPlugin from 'remark-html';
import breaks from 'remark-breaks';

interface OwnProps {}
interface StateProps {
  title: string;
  markdown: string;
}
interface DispatchProps {
  changeMarkdown: typeof changeMarkdown;
  changeTitle: typeof changeTitle;
  setHtml: typeof setHtml;
  convertEditorMode: typeof convertEditorMode;
}
type MarkdownEditorContainerProps = OwnProps & StateProps & DispatchProps;

const MarkdownEditorContainer: React.SFC<MarkdownEditorContainerProps> = ({
  changeMarkdown,
  changeTitle,
  setHtml,
  convertEditorMode,
  title,
  markdown,
}) => {
  const onConvert = (markdown: string) => {
    remark()
      .use(breaks)
      .use(htmlPlugin)
      .process(markdown, (err: any, file: any) => {
        const html = String(file);
        setHtml(html);
        convertEditorMode();
      });
  };
  return (
    <MarkdownEditor
      title={title}
      markdown={markdown}
      onChangeMarkdown={changeMarkdown}
      onChangeTitle={changeTitle}
      onConvert={onConvert}
    />
  );
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  state => ({
    title: state.write.title,
    markdown: state.write.markdown,
  }),
  {
    changeMarkdown,
    changeTitle,
    setHtml,
    convertEditorMode,
  },
)(MarkdownEditorContainer);
