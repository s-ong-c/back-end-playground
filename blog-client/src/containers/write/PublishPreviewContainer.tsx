import * as React from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import PublishPreview from '../../components/write/PublishPreview';

const mapStateToProps = (state: RootState) => ({
  title: state.write.title,
});
const mapDispatchToProps = {};

interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
export type PublishPreviewContainerProps = OwnProps &
  StateProps &
  DispatchProps;

const PublishPreviewContainer: React.SFC<PublishPreviewContainerProps> = ({
  title,
}) => {
  return <PublishPreview title={title} />;
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(PublishPreviewContainer);
