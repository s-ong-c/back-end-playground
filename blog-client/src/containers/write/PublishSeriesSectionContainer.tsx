import * as React from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import PublishSeriesSection from '../../components/write/PublishSeriesSection';

interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
type PublishSeriesSectionContainerProps = OwnProps & StateProps & DispatchProps;

const mapStateToProps = (state: RootState) => ({});
const mapDispatchToProps = {};

const PublishSeriesSectionContainer: React.SFC<
  PublishSeriesSectionContainerProps
> = props => {
  return <PublishSeriesSection />;
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(PublishSeriesSectionContainer);
