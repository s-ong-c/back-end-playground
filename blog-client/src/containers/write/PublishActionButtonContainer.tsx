import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '../../modules';
import PublishActionButton from '../../components/write/PublishActionButton';
import { closePublish, WriteMode } from '../../modules/write';
import { WRITE_POST, WritePostResponse } from '../../lib/graphql/post';
import { pick } from 'ramda';
import { escapeForUrl } from '../../lib/utils';
import { useMutation, useApolloClient } from '@apollo/react-hooks';
import { setHeadingId } from '../../lib/heading';
import { useHistory } from 'react-router';

// const mapStateToProps = ({ write }: RootState) =>
//   pick(
//     [
//       'mode',
//       'markdown',
//       'title',
//       'html',
//       'tags',
//       'defaultDescription',
//       'description',
//       'isPrivate',
//       'urlSlug',
//       'thumbnail',
//     ],
//     write,
//   );
// const mapDispatchToProps = {
//   closePublish,
// };

// interface OwnProps {}
// type StateProps = ReturnType<typeof mapStateToProps>;
// type DispatchProps = typeof mapDispatchToProps;
type PublishActionButtonsContainerProps = {};

const PublishActionButtonsContainer: React.FC<PublishActionButtonsContainerProps> = props => {
  const dispatch = useDispatch();
  const client = useApolloClient();
  const history = useHistory();

  const options = useSelector((state: RootState) =>
    pick(
      [
        'mode',
        'markdown',
        'title',
        'html',
        'tags',
        'defaultDescription',
        'description',
        'isPrivate',
        'urlSlug',
        'thumbnail',
      ],
      state.write,
    ),
  );
  const [writePost] = useMutation<WritePostResponse>(WRITE_POST);

  const variables = {
    title: options.title,
    body:
      options.mode === WriteMode.MARKDOWN
        ? options.markdown
        : setHeadingId(options.html),
    tags: options.tags,
    is_markdown: options.mode === WriteMode.MARKDOWN,
    is_temp: false,
    is_private: options.isPrivate,
    url_slug: options.urlSlug || escapeForUrl(options.title),
    thumbnail: options.thumbnail,
    meta: {
      short_description: options.description,
    },
  };

  const onPublish = async () => {
    if (options.title.trim() === '') {
      console.error('제목이 비어있습니다.');
      return;
    }
    try {
      const response = await writePost({
        variables: variables,
      });
      if (!response || !response.data) return;
      const { user, url_slug } = response.data.writePost;
      await client.resetStore();
      history.push(`/@${user.username}/${url_slug}`);
    } catch (e) {
      console.error('포스트 작성 실패');
    }
  };

  const onCancel = useCallback(() => {
    dispatch(closePublish());
  }, [dispatch]);

  return <PublishActionButton onCancel={onCancel} onPublish={onPublish} />;
  // return (
  //   <Mutation mutation={WRITE_POST}>
  //     {(
  //       writePost: (arg0: {
  //         variables: {
  //           title: string;
  //           body: string;
  //           tags: string[];
  //           is_markdown: boolean;
  //           is_temp: boolean;
  //           is_private: boolean;
  //           url_slug: string;
  //           thumbnail: string | null;
  //           meta: {};
  //         };
  //       }) => void,
  //       { data, loading, error }: MutationResult<Partial<Post>>,
  //     ) => {
  //       return (
  //         <PublishActionButton
  //           onCancel={onCancel}
  //           onPublish={async () => {
  //             try {
  //               const response = await writePost({
  //                 variables: {
  //                   title: rest.title,
  //                   body:
  //                     rest.mode === WriteMode.MARKDOWN
  //                       ? rest.markdown
  //                       : rest.html,
  //                   tags: rest.tags,
  //                   is_markdown: rest.mode === WriteMode.MARKDOWN,
  //                   is_temp: false,
  //                   is_private: rest.isPrivate,
  //                   url_slug: rest.urlSlug || escapeForUrl(rest.title),
  //                   thumbnail: rest.thumbnail,
  //                   meta: {},
  //                 },
  //               });
  //               console.log(response);
  //             } catch (e) {
  //               console.log(e);
  //             }
  //           }}
  //         />
  //       );
  //     }}
  //   </Mutation>
};

export default PublishActionButtonsContainer;
