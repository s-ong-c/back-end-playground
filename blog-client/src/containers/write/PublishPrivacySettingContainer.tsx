import * as React from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import PublishPrivacySettings from '../../components/write/PublishPrivacySettings';
import { setPrivacy } from '../../modules/write';
const mapStateToProps = (state: RootState) => ({
  isPrivate: state.write.isPrivate,
});
const mapDispatchToProps = {
  setPrivacy,
};
interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
export type PublishPrivacySettingContainerProps = OwnProps &
  StateProps &
  DispatchProps;

const PublishPrivacySettingContainer: React.SFC<
  PublishPrivacySettingContainerProps
> = ({ isPrivate, setPrivacy }) => {
  const onSelect = React.useCallback(
    isPrivate => {
      setPrivacy(isPrivate);
    },
    [setPrivacy],
  );
  return <PublishPrivacySettings isPrivate={isPrivate} onSelect={onSelect} />;
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(PublishPrivacySettingContainer);
