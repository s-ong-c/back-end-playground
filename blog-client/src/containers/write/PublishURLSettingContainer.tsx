import React, { useCallback } from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import PublishURLSetting from '../../components/write/PublishURLSetting';
import { escapeForUrl } from '../../lib/utils';
import { changeUrlSlug } from '../../modules/write';

const mapStateToProps = (state: RootState) => ({
  username: state.core.user && state.core.user.username,
  title: state.write.title,
  urlSlug: state.write.urlSlug,
});
const mapDispatchToProps = {
  changeUrlSlug,
};
interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
export type PublishURLSettingContainerProps = OwnProps &
  StateProps &
  DispatchProps;

const PublishURLSettingContainer: React.FC<PublishURLSettingContainerProps> = ({
  username,
  title,
  urlSlug,
  changeUrlSlug,
}) => {
  const defaultUrlSlug = escapeForUrl(title);
  const urlSlugToShow = urlSlug || defaultUrlSlug;
  const onChangeUrlSlug = useCallback(urlSlug => changeUrlSlug(urlSlug), [
    changeUrlSlug,
  ]);
  if (!username) return null;
  return (
    <PublishURLSetting
      username={username}
      urlSlug={urlSlugToShow}
      onChangeUrlSlug={onChangeUrlSlug}
    />
  );
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(PublishURLSettingContainer);
