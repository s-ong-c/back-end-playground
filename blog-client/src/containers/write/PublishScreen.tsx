import * as React from 'react';
import { connect } from 'react-redux';
import { RootState } from '../../modules';
import PublishScreenTemplate from '../../components/write/PublishScreenTemplate';
import PublishPreviewContainer from './PublishPreviewContainer';
import PublishPrivacySettingContainer from './PublishPrivacySettingContainer';
import PublishURLSettingContainer from './PublishURLSettingContainer';
import PublishSeriesSectionContainer from './PublishSeriesSectionContainer';
import PublishActionButton from '../../components/write/PublishActionButton';

interface OwnProps {}
type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;
type PublishScreenProps = OwnProps & StateProps & DispatchProps;

const mapStateToProps = (state: RootState) => ({});
const mapDispatchToProps = {};

const PublishScreen: React.SFC<PublishScreenProps> = props => {
  return (
    <PublishScreenTemplate
      left={<PublishPreviewContainer />}
      right={
        <>
          <div>
            <PublishPrivacySettingContainer />
            <PublishURLSettingContainer />
            <PublishSeriesSectionContainer />
          </div>
          <PublishActionButton />
        </>
      }
    />
  );
};

export default connect<StateProps, DispatchProps, OwnProps, RootState>(
  mapStateToProps,
  mapDispatchToProps,
)(PublishScreen);
