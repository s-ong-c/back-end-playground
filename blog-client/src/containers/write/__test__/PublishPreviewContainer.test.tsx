import * as React from 'react';
import { render } from 'react-testing-library';
import PublishPreviewContainer, {
  PublishPreviewContainerProps,
} from '../PublishPreviewContainer';
import { Provider } from 'react-redux';
import rootReducer from '../../../modules';
import { createStore } from 'redux';
import { changeTitle } from '../../../modules/write';

describe('PublishPreview', () => {
  const setup = (props: Partial<PublishPreviewContainerProps> = {}) => {
    const store = createStore(rootReducer);
    const utils = render(
      <Provider store={store}>
        <PublishPreviewContainer {...props} />
      </Provider>,
    );
    return {
      ...utils,
      store,
    };
  };
  it('renders property', () => {
    setup();
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  it('renders title', () => {
    const utils = setup();
    utils.store.dispatch(changeTitle('title is changed'));
    utils.getByText('title is changed');
  });
});
