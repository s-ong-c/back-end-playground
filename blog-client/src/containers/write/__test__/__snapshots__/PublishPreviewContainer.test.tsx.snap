// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublishPreview matches snapshot 1`] = `
<div>
  <section
    class="sc-bwzfXH uTxCW sc-bdVaJa jSADnk"
  >
    <h3>
      포스트카드 미리보기
    </h3>
    <div
      class="contents"
    >
      <div
        class="sc-htpNat fGfspC"
      >
        <div
          class="sc-bxivhb oVXVi"
        >
          <div
            class="sc-ifAKCX bnjEss"
          >
            <svg>
              vector-image.svg
            </svg>
            <button
              class="sc-EHOje gyfGXD"
            >
              썸네일 업로드
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-bZQynM dcFljb"
      >
        <h4 />
        <textarea
          class="sc-gzVnrw eOJEaf"
        />
        <div
          class="sc-htoDjs dTqKOk"
        >
          118/150
        </div>
      </div>
    </div>
  </section>
</div>
`;
